pypsa/linopt.py

```diff
@@ -601,6 +601,175 @@ def set_int_index(ser):
+    """
+    options_fn = "highs_options.txt"
+    default_dict = {
+        "method": "ipm",
+        "primal_feasibility_tolerance": 1e-04,
+        "dual_feasibility_tolerance": 1e-05,
+        "ipm_optimality_tolerance": 1e-6,
+        # "ipm_iteration_limit": 100,
+        "presolve": "on",
+        "run_crossover": True,
+        "parallel": "off",
+        "threads": 4,
+        "solution_file": solution_fn,
+        "write_solution_to_file": True,
+        "write_solution_style": 1,
+        "log_to_console": True,
+    }
+    # update default_dict through solver_options and write to file
+    default_dict.update(solver_options)
+    method = default_dict.pop("method", "ipm")
+    logger.info(f"Options: \"{default_dict}\". List of options: https://www.maths.ed.ac.uk/hall/HiGHS/HighsOptions.set")
+    f1 = open(options_fn, "w")
+    f1.write('\n'.join([f"{k} = {v}" for k, v in default_dict.items()]))
+    f1.close()
+
+    # write (terminal) commands
+    command = f"highs --model_file {problem_fn} "
+    if warmstart:
+        logger.warning("Warmstart, not available in HiGHS. Will be ignored.")
+    command += f"--solver {method} --options_file {options_fn}"
+    logger.info(f"Solver command: \"{command}\"")
+    # execute command and store command window output
+    process = subprocess.Popen(
+        command.split(' '),
+        stdout=subprocess.PIPE,
+        universal_newlines=True
+    )
+
+    def read_until_break():
+        # Function that reads line by line the command window
+        while True:
+            out = process.stdout.readline(1)
+            if out == '' and process.poll() != None:
+                break
+            if out != '':
+                yield out
+
+    # converts stdout (standard terminal output) to pandas dataframe
+    log = io.StringIO(''.join(read_until_break())[:])
+    log = pd.read_csv(log, sep=':', index_col=0, header=None)[1].squeeze()
+    if solver_logfile is not None:
+        log.to_csv(solver_logfile, sep="\t")
+    log.index = log.index.str.strip()
+    os.remove(options_fn)
+
+    # read out termination_condition from `info`
+    model_status = log["Model   status"].strip().lower()
+    if "optimal" in model_status:
+        status = "ok"
+        termination_condition = model_status
+    elif "infeasible" in model_status:
+        status = "warning"
+        termination_condition = model_status
+    else:
+        status = 'warning'
+        termination_condition = model_status
+    objective = float(log["Objective value"])
+
+    # read out solution file (.sol)
+    f = open(solution_fn, "rb")
+    trimed_sol_fn = re.sub(rb'\*\*\s+', b'', f.read())
+    f.close()
+    
+    sol = pd.read_csv(io.BytesIO(trimed_sol_fn), header=[1], sep=r'\s+')
+    row_no = sol[sol["Index"] == 'Rows'].index[0]
+    sol = sol.drop(row_no+1)  # Removes header line after "Rows"
+    sol_rows = sol[(sol.index > row_no)]
+    sol_cols = sol[(sol.index < row_no)].set_index("Name").pipe(set_int_index)
```

Comment:
Wouldn't it be possible to use `set_int_index` as well for `sol_rows`. Then you wouldn't need the `reset_index(drop=True) + 1` below.

Reply:
This does not work. Below is the dataframe of `sol_rows`. While sol_cols have variables with a "x" i.e. x300, `sol_rows` doesn't have that. Therefore, set_int_index does not work (could be modified).

![image](https://user-images.githubusercontent.com/61968949/142735951-65b306fd-28fb-468c-b15c-dd8522233190.png)

Linopy will be added soon, I think it's probably not useful to modify the set_int_index function?

Reply:
That sounds absolutely reasonable. Thought the duals may have a "y" prefix.